### PR TITLE
Using BufferedStream over NetworkStream for performance improvement.

### PR DESCRIPTION
**Pull request by navin22**
- Used BufferedStream over NetworkStream while writing the data to reduce the number of calls.
- Used BufferedSize of 16KB.
- Didn't use BufferedStream while reading, as this didn't add much perf improvements. And was hanging when we were doing Socket.Poll.

### CODE REVIEW

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketCommunicationManager.cs [hunk 1]
```diff
@@ -119,12 +118,13 @@ public IPEndPoint HostServer(IPEndPoint endpoint)
 
                 var client = await this.tcpListener.AcceptTcpClientAsync();
                 this.socket = client.Client;
-                this.stream = client.GetStream();
-                this.binaryReader = new BinaryReader(this.stream);
+                this.socket.NoDelay = true;
+                this.stream = new PlatformStream().PlatformBufferedStream(client.GetStream(), SocketConstants.BUFFERSIZE);
+                this.binaryReader = new BinaryReader(client.GetStream());
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/SocketCommunicationManager.cs]: Is it possible to add a test for this case in Platform Tests suite?

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketCommunicationManager.cs [hunk 2]
```diff
@@ -354,8 +355,8 @@ private string TryReceiveRawMessage(CancellationToken cancellationToken)
                         && socketException.SocketErrorCode == SocketError.TimedOut)
                     {
                         EqtTrace.Info(
-                            "SocketCommunicationManager ReceiveMessage: failed to receive message because read timeout {0}",
-                            ioException);
+                        "SocketCommunicationManager ReceiveMessage: failed to receive message because read timeout {0}",
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/SocketCommunicationManager.cs]: nit: not required change?

File: src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/System/IStream.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/System/IStream.cs]: Suggest `Interfaces/IO` for this. It is independent of system characteristics.

File: src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/System/IStream.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    /// Helper class to return plaform specific stream.
+    /// </summary>
+    public interface IStream
+    {
+        /// <summary>
+        /// Returns plarform specific Buffered Stream
+        /// </summary>
+        /// <param name="stream">Input Stream</param>
+        /// <returns>Buffered Stream</returns>
+        Stream PlatformBufferedStream(Stream stream);
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/System/IStream.cs]: Don't see usage of this API. Can we remove it please?

File: src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/System/IStream.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+        /// <summary>
+        /// Returns platrform specific Buffered Stream with desired buffer size.
+        /// </summary>
+        /// <param name="stream">Input Stream</param>
+        /// <param name="bufferSize">Buffer Size</param>
+        /// <returns>Buffered Stream</returns>
+        Stream PlatformBufferedStream(Stream stream, int bufferSize);
```
Comment (codito) [inline src/Microsoft.TestPlatform.PlatformAbstractions/Interfaces/System/IStream.cs]: Could be `CreateBufferedStream`.

Or we could make it an extension method on `Stream`. E.g. `Stream.ToBufferedStream()`.

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketCommunicationManager.cs (hunk 2 above)
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketCommunicationManager.cs]: Hey, It was already there. 


---
In reply to: [137164236](https://github.com/Microsoft/vstest/pull/1041#discussion_r137164236) [](ancestors = 137164236)

Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketCommunicationManager.cs]: Yeah, added the tabs back.

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs [hunk 6]
```diff
@@ -69,7 +70,7 @@ private void OnServerConnected(Task connectAsyncTask)
                 throw connectAsyncTask.Exception;
             }
 
-            this.channel = this.channelFactory(this.tcpClient.GetStream());
+            this.channel = this.channelFactory(new PlatformStream().PlatformBufferedStream(this.tcpClient.GetStream(), SocketConstants.BUFFERSIZE));
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: Seems like we are using `BufferedStream` for both Read and Write operations. We need to change it to comply with recommended approach.

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs [hunk 7]
```diff
@@ -35,7 +36,7 @@ protected SocketClient(Func<Stream, ICommunicationChannel> channelFactory)
             this.cancellation = new CancellationTokenSource();
             this.stopped = false;
 
-            this.tcpClient = new TcpClient();
+            this.tcpClient = new TcpClient { NoDelay = true };
```
Comment (harshjain2) [inline src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: Using `NoDelay = true` can be a bit inefficient. 
More info here :
https://msdn.microsoft.com/en-us/library/system.net.sockets.tcpclient.nodelay(v=vs.110).aspx

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs (hunk 6 above)
Comment (harshjain2) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: Same applies for SocketServer.cs

File: test/Microsoft.TestPlatform.PerformanceTests/SocketTests.cs [hunk 8]
```diff
@@ -6,6 +6,8 @@ namespace Microsoft.TestPlatform.PerformanceTests
     using System;
     using System.Diagnostics;
     using System.Net;
+    using System.Collections.Generic;
```
Comment (harshjain2) [inline test/Microsoft.TestPlatform.PerformanceTests/SocketTests.cs]: nit : ordering of namespaces.

**Review by harshjain2 — CHANGES_REQUESTED** [review]

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketCommunicationManager.cs [hunk 9]
```diff
@@ -119,12 +118,13 @@ public IPEndPoint HostServer(IPEndPoint endpoint)
 
                 var client = await this.tcpListener.AcceptTcpClientAsync();
                 this.socket = client.Client;
-                this.stream = client.GetStream();
-                this.binaryReader = new BinaryReader(this.stream);
+                this.socket.NoDelay = true;
+                this.stream = new PlatformStream().PlatformBufferedStream(client.GetStream(), SocketConstants.BUFFERSIZE);
```
Comment (smadala) [inline src/Microsoft.TestPlatform.CommunicationUtilities/SocketCommunicationManager.cs]: Add comment.

Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketCommunicationManager.cs]: Got it, as we do socket.poll here also. Need to do in Socket.Client and Socket.Server.

---
In reply to: [137203711](https://github.com/Microsoft/vstest/pull/1041#discussion_r137203711) [](ancestors = 137203711)

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs (hunk 7 above)
Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketClient.cs]: Yeah, As it also says we need for immediate/important communications. 

---
In reply to: [137197765](https://github.com/Microsoft/vstest/pull/1041#discussion_r137197765) [](ancestors = 137197765)

**Review by harshjain2 — APPROVED** [review]

File: src/Microsoft.TestPlatform.CommunicationUtilities/LengthPrefixCommunicationChannel.cs [hunk 10]
```diff
@@ -27,7 +27,9 @@ public LengthPrefixCommunicationChannel(Stream stream)
         {
             this.stream = stream;
             this.reader = new BinaryReader(stream, Encoding.UTF8, true);
-            this.writer = new BinaryWriter(stream, Encoding.UTF8, true);
+
+            // Using the Buffered stream while writing.
```
Comment (codito) [inline src/Microsoft.TestPlatform.CommunicationUtilities/LengthPrefixCommunicationChannel.cs]: Good to specify why.

File: test/Microsoft.TestPlatform.PerformanceTests/SocketTests.cs [hunk 11]
```diff
@@ -4,8 +4,10 @@
 namespace Microsoft.TestPlatform.PerformanceTests
 {
     using System;
+    using System.Collections.Generic;
```
Comment (codito) [inline test/Microsoft.TestPlatform.PerformanceTests/SocketTests.cs]: These changes are not required?

**Review by codito — APPROVED** [review]
Looks fantastic. Suggest few nit changes.

File: src/Microsoft.TestPlatform.CommunicationUtilities/SocketConstants.cs [hunk 12]
```diff
@@ -0,0 +1,11 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities
+{
+    public class SocketConstants
+    {
+        // Buffer size for the buffered stream we are using.
+        public const int BUFFERSIZE = 16384;
```
Comment (IanKemp) [inline src/Microsoft.TestPlatform.CommunicationUtilities/SocketConstants.cs]: Nit: Isn't the MS standard to use PascalCase for constant names?

Comment (navin22) [reply src/Microsoft.TestPlatform.CommunicationUtilities/SocketConstants.cs]: Yeah, Thanks !

---
In reply to: [137451610](https://github.com/Microsoft/vstest/pull/1041#discussion_r137451610) [](ancestors = 137451610)

**Pull request merged**